api/agent/protocol/json.go

```diff
@@ -35,54 +46,92 @@ func writeString(err error, dst io.Writer, str string) error {
 
 // TODO(xxx): headers, query parameters, body - what else should we add to func's payload?
 // TODO(xxx): get rid of request body buffering somehow
-func (h *JSONProtocol) DumpJSON(req *http.Request) error {
+func (h *JSONProtocol) DumpJSON(call *models.Call, req *http.Request) error {
 	stdin := json.NewEncoder(h.in)
 	bb := new(bytes.Buffer)
 	_, err := bb.ReadFrom(req.Body)
+	// todo: better/simpler err handling
 	if err != nil {
 		return err
 	}
+	// open
 	err = writeString(err, h.in, "{")
+	if err != nil {
+		return err
+	}
+
+	// call_id
+	err = writeString(err, h.in, `"call_id":`)
+	if err != nil {
+		return err
+	}
+	err = stdin.Encode(call.ID)
+	if err != nil {
+		return err
+	}
+
+	// body
+	err = writeString(err, h.in, ",")
 	err = writeString(err, h.in, `"body":`)
 	if err != nil {
 		return err
 	}
 	err = stdin.Encode(bb.String())
+	if err != nil {
+		return err
+	}
+
+	// request URL
 	err = writeString(err, h.in, ",")
-	err = writeString(err, h.in, `"headers":`)
+	err = writeString(err, h.in, `"request_url":`)
 	if err != nil {
 		return err
 	}
-	err = stdin.Encode(req.Header)
+	err = stdin.Encode(req.URL.String())
+	if err != nil {
+		return err
+	}
+
+	// headers
 	err = writeString(err, h.in, ",")
-	err = writeString(err, h.in, `"query_parameters":`)
+	err = writeString(err, h.in, `"headers":`)
 	if err != nil {
 		return err
 	}
-	err = stdin.Encode(req.URL.RawQuery)
-	err = writeString(err, h.in, "}")
+	err = stdin.Encode(req.Header)
+
+	// config
+	if call.EnvVars != nil && len(call.EnvVars) > 0 {
+		err = writeString(err, h.in, ",")
+		err = writeString(err, h.in, `"config":`)
+		if err != nil {
+			return err
+		}
+		err = stdin.Encode(call.EnvVars)
+	}
+
+	// close
+	err = writeString(err, h.in, "}\n\n")
```

Comment:
why do we need to write `}\n\n`

If you are reading JSON payload byte-by-byte then you should do following:
```
    while True: # read from the stream until some of the following checks will break it
        c = stream.read(1)
        if c is None:
            continue # this is actual emulates blocking, and attempts to read one more time again and again.
        if len(c) == 0: # means that previous byte was the last in payload.
            return json(s)
        if c == "\n": # this one follows the way we write payload and expects that next byte will be the last before exiting loop
            line += c
            ret = True
        elif c == "}" and ret: # this is the last byte handler
            line += c
            return json(s)
        else: # if not - keep reading until next matched check.
            ret = False
            line += c
```
So, i'm not kinda agreed to do add more bytes payload.

Reply:
Wasn't clear how it was being delineated, now it's very clear:

```
...OBJECT 1
}\n
\n
{
OBJECT 2...
```